### PR TITLE
Prefs [TASK][28591] Hotel Presenter (Hotel Tab)

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 28591: Hotel Presenter(Hotel Tab)](https://edu-redmine.sun-asterisk.vn/issues/28591)

## WHAT
- Hotel Presenter(Hotel Tab)
## HOW
- Hotel Presenter(Hotel Tab)
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    init {
+        View.inflate(context, R.layout.layout_traveler_pick, this)
+        max = resources.getInteger(R.integer.integer_10)
+        min = resources.getInteger(R.integer.integer_0)
+        numberChildren = DestinationFilter.childrenCount
+        numberAdult = DestinationFilter.adultCount
+        textViewPickAdult.text = numberAdult.toString()
+        textViewPickChildren.text = numberChildren.toString()
+        handleEvents()
+    }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt]:
```suggestion
    private var picker: TextView? = null
    private var numberChildren: Int = DestinationFilter.childrenCount
    private var numberAdult: Int = DestinationFilter.adultCount
    private var dialogFragment: DialogFragment? = null
    private var max: Int = resources.getInteger(R.integer.integer_10)
    private var min: Int = resources.getInteger(R.integer.integer_0)

    init {
        View.inflate(context, R.layout.layout_traveler_pick, this)
        textViewPickAdult.text = numberAdult.toString()
        textViewPickChildren.text = numberChildren.toString()
        handleEvents()
    }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    private fun increaseChildren() {
+        when (numberChildren) {
+            in min..max -> {
+                numberChildren++
+                if (numberChildren > max) { numberChildren = max }
+                if (numberChildren < min) { numberChildren = min }
+                textViewPickChildren.text = numberChildren.toString()
+            }
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt]: Chuyển sang `if`

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    private fun decreaseChildren() {
+        when (numberChildren) {
+            in min..max -> {
+                numberChildren--
+                if (numberChildren > max) { numberChildren = max }
+                if (numberChildren < min) { numberChildren = min }
+                textViewPickChildren.text = numberChildren.toString()
+            }
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt]: chuyển sang `if`
```suggestion
        if (numberChildre !in min..max) return
        numberChildren--
        if (numberChildren > max) numberChildren = max
        if (numberChildren < min) numberChildren = min
        textViewPickChildren.text = numberChildren.toString()
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    private fun increaseAdult() {
+        when (numberAdult) {
+            in min..max -> {
+                numberAdult++
+                if (numberAdult > max) { numberAdult = max }
+                if (numberAdult < min) { numberAdult = min }
+                textViewPickAdult.text = numberAdult.toString()
+            }
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt]: Chuyển sang `if`

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    private fun decreaseAdult() {
+        when (numberAdult) {
+            in min..max -> {
+                numberAdult--
+                if (numberAdult > max) { numberAdult = max }
+                if (numberAdult < min) { numberAdult = min }
+                textViewPickAdult.text = numberAdult.toString()
+            }
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt]: Chuyển sang `if`

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelContract.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.sunasterisk.travelapp.ui.home.hotel
+
+import com.sunasterisk.travelapp.data.models.Location
+import com.sunasterisk.travelapp.ui.base.BaseContract
+
+interface HotelContract {
+
+    interface View : BaseContract.View<Presenter> {
+        fun updateLocations(list: List<Location>)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelContract.kt]:
```suggestion
        fun updateLocations(locations: List<Location>)
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                override fun onSuccess(data: List<Location>) {
+                    view?.updateLocations(data)
+                    view?.dismissProgressDialog()
+                }
+
+                override fun onError(throwable: Throwable) {
+                    view?.onError(throwable.message)
+                    view?.dismissProgressDialog()
+                }
+            })
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelPresenter.kt]: Cho `dismissProgressDialog` vào trong hàm `updateLocation` và `onError` của view

**Review by aug16vcc — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelPresenter.kt (hunk 7 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelPresenter.kt]: ok ong

**Review by aug16vcc — APPROVED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        textViewPickAdult.text = numberAdult.toString()
+    }
+
+    fun setPicker(picker: TextView) {
+        this.picker = picker
+    }
+
+    fun setDialogFragment(dialogFragment: DialogFragment) {
+        this.dialogFragment = dialogFragment
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt:106]: class layout 
public
private

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        if (numberAdult < min) numberAdult = min
+        textViewPickAdult.text = numberAdult.toString()
+    }
+
+    private fun decreaseAdult() {
+        if (numberAdult !in min..max) return
+        numberAdult--
+        if (numberAdult > max) numberAdult = max
+        if (numberAdult < min) numberAdult = min
+        textViewPickAdult.text = numberAdult.toString()
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt:97]: Tại sao number-- lại phải so sánh với max ? 
đoạn này logic là gì thế

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import android.util.AttributeSet
+import android.view.View
+import android.widget.TextView
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.fragment.app.DialogFragment
+import com.sunasterisk.travelapp.R
+import com.sunasterisk.travelapp.data.models.DestinationFilter
+import kotlinx.android.synthetic.main.layout_traveler_pick.view.*
+
+class PeopleNumberPick @JvmOverloads constructor(
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/PeopleNumberPick.kt:13]:
```suggestion
class PeopleNumberPick(
```

File: app/src/main/res/layout/fragment_pick_traveler.xml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<com.sunasterisk.travelapp.ui.custom.PeopleNumberPick
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/peopleNumberPick"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+</com.sunasterisk.travelapp.ui.custom.PeopleNumberPick>
```
Comment (huytq-2422) [inline app/src/main/res/layout/fragment_pick_traveler.xml:7]: .

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 12]
```diff
@@ -71,6 +88,28 @@ class HotelTabFragment private constructor() : BaseFragment(), View.OnClickListe
...
+                textViewFilterTravelers.post {
+                    dialog.setPickView(textViewFilterTravelers)
+                }
+            }
+        }
+    }
+
+    override fun updateLocations(list: List<Location>) {
+        adapter.updateData(list)
+        dismissProgressDialog()
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt:110]: update Locations thì có liên quan gì đến dismiss Progress Dialog đâu 
Vi phạm nguyên lý đầu tiên của SOLID nhá

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 13]
```diff
@@ -85,25 +124,16 @@ class HotelTabFragment private constructor() : BaseFragment(), View.OnClickListe
...
-                    dialog.show(
-                        manager,
-                        TravelersPickDialogFragment::class.simpleName
-                    )
-                }
-            }
-        }
+    override fun onError(message: String?) {
+        super.onError(message)
+        dismissProgressDialog()
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt:137]: tương tự onError cũng k liên quan đến việc dismiss Dialog

**Pull request closed without merging**